test/widgets/test_notify.py

```diff
@@ -0,0 +1,216 @@
+# Copyright (c) 2021 elParaguayo
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# NOTE: This test only tests the functionality of the widget and parts of the manager
+# The notification service (in libqtile/notify.py) is tested separately
+# TO DO: notification service test ;)
+
+import sys
+from importlib import reload
+from types import ModuleType
+
+import pytest
+
+import libqtile.bar
+import libqtile.config
+from libqtile.widget import notify
+
+
+# We need to fake dbus_next modules so the NotificationManager can be imported
+class MockDBNService(ModuleType):
+    class ServiceInterface:
+        pass
+
+    # decorators used in NotificationService
+    def method(cls):  # noqa: N805
+        def wrap(func):
+            return func
+        return wrap
+
+    def signal(cls):  # noqa: N805
+        def wrap(func):
+            return func
+        return wrap
+
+
+class MockDBNAio(ModuleType):
+    class MessageBus:
+        pass
+
+
+# Tell the manager that the service is running
+async def mock_service():
+    return True
+
+
+# Bit of a hack... when we log a timer, save the delay in an attribute
+# We'll use this to check message timeouts are being honoured.
+def log_timeout(self, delay, func):
+    self.delay = delay
+    self.qtile.call_later(delay, func)
+
+
+# Monkeypatch dbus_next modules and import notify module
+@pytest.fixture
+def notify_utility(monkeypatch):
+    monkeypatch.setitem(sys.modules, "dbus_next.aio", MockDBNAio("dbus_next.aio"))
+    monkeypatch.setitem(sys.modules, "dbus_next.service", MockDBNService("dbus_next.service"))
+    import libqtile.notify
+    reload(libqtile.notify)
+    return libqtile.notify
+
+
+# To send notifications via the manager object we need to send a string to eval
```

Comment:
Why can't we just send the notifications with dbus?

Reply:
That would certainly be neater... However, what tools are available to send notifications with dbus? I think we've excluded `dbus_next` from the building environment. What did you have in mind?

Reply:
notify-send or alike? But I guess isolation is the next problem, as we don't want to spam the host machine with notifications.

Reply:
We should probably just write a helper function that makes our notification server think there is a new notification coming in.

Reply:
Ah, but our notification service only exist if `dbus_next` is installed.

> I think we've excluded dbus_next from the building environment as it's only an optional dependency.

What is the rationale?

Worst case scenario, we could probably mock a notification manager, if I understand the code in `notify.py` correctly. That would be much cleaner than here, I think.

Reply:
If you can't isolate then the problem works the other way too e.g. if some other process on the host machine sends a notification when we're running our test, we'll get unexpected results.

Reply:
Yeah right. :)

What do you think of going with a custom notification manager then?

Reply:
I did think of that as an approach to use here and decided against it as I wanted to use the existing manager,

However, this test should really just check that the widget displayed notifications received from the attached manager so no reason not to use a custom one. It would probably make this one easier to read!

I was planning to do a separate test for the `NotificationService` and `NotificationManager` (not sure how, yet...).

Reply:
Would it not be strictly superior as a test to install `dbus-next` and run `dbus` proper, establish the notification server and send it a real notification?

Reply:
Depends... do you want to test everything (service, manager and widget) in one test or do you test them separately?

Also, I don't know how the testing environment works. Is the notification service isolated from other processes (not just our various tests but any other processes on the server)?

Reply:
I think test them separately or together where appropriate. I'm assuming that installing dbus and starting it with `dbus-launch` (https://dbus.freedesktop.org/doc/dbus-launch.1.html) before starting the tests would be enough, but I can't say I'm super familiar with it so not sure. I think you can get the dbus address from dbus-launch as it starts up and stick that in the environmental variables so that any dbus-interfacing code knows what to do. I guess `dbus-next` will "just work" as normal if it can get the address from the environment.

Reply:
Interesting. Let me do some investigating. As you say, `dbus_next` looks up the bus address from the environment so this may work.

Reply:
Any suggestions as to how I could test this locally (i.e. setting up a minimal environment and running dbus inside it etc?).

Alternative is likely to be lots of different commits being pushed and waiting for CI to run all the tests (the widgets are done last!)

Reply:
I think something kinda like this might do it (there is probably a neater way):
```diff
diff --git a/test/conftest.py b/test/conftest.py
index 76c49923..8ce88410 100644
--- a/test/conftest.py
+++ b/test/conftest.py
@@ -487,6 +487,27 @@ def xephyr(request, xvfb):  # noqa: F841
         yield x
 
 
+@pytest.fixture(scope="function")
+def dbus(request):  # noqa: F841
+    use = getattr(request, "param", False)
+    if use:
+        launcher = whereis("dbus-launch")
+        if launcher is None:
+            yield None  # should probably raise an exception
+        else:
+            result = subprocess.run(launcher, stdout=subprocess.PIPE)
+            if result.returncode or not result.stdout:
+                yield None  # likewise
+            else:
+                env = {}
+                for line in result.stdout.strip():
+                    var, val = line.split("=")
+                    env[var] = val
+                yield env
+    else:
+        yield None
+
+
 @pytest.fixture(scope="function")
 def manager(request, xephyr):
     config = getattr(request, "param", BareConfig)
```
This can be parametrised with `True` to enable it for a test. I guess it should probably be used by `manager` or maybe even `xephyr` rather than tests directly so that the `env` that it yields can be put in the right place where downstream code (i.e. `dbus-next` can find the right environmental variables. I think that should work fine.

Reply:
Thanks. I was more concerned about testing locally (where I have `dbus_next` and `dbus` installed already) and being able to check whether this solution works before pushing to GitHub for the CI checks. i.e. just because a test works on my local machine, it doesn't guarantee it will work in CI.

Reply:
I see... can `dbus_next` use dbus even if the environmental has been cleared of dbus-related variables?

Reply:
> I was planning to do a separate test for the `NotificationService` and `NotificationManager` (not sure how, yet...).

This is probably what we want to start with here. :)

Can you remove the notifications part of this PR so that I can merge the rest?

Reply:
Done.

Reply:
Update for the both of you: I stripped down `ci.yml` and `tox.ini` on a fork on my clone so that it only runs the `test_notify` and removes all other stuff to keep the testing really quick. I added some new dependencies and I can the widget to test successfully using `dbus_next`.

Lots more to do before I can submit a PR but it looks promising.

Reply:
Awesome!